lib/restforce/db/runner_cache.rb

```diff
@@ -0,0 +1,90 @@
+module Restforce
+
+  module DB
+
+    # Restforce::DB::RunnerCache serves as a means of caching the collections of
+    # recently-updated database and Salesforce instances for passed mappings.
+    # The general goal is to avoid making repetitive Salesforce API calls or
+    # database queries, and ensure a consistent list of objects during a
+    # synchronization run.
+    class RunnerCache
+
+      # Public: Initialize a new Restforce::DB::RunnerCache.
+      def initialize
+        reset
+      end
+
+      # Public: Iterate through the recently-updated instances of the specified
+      # type for the passed mapping. Memoizes the records if the collection has
+      # not previously been cached.
+      #
+      # mapping     - A Restforce::DB::Mapping.
+      # record_type - A Symbol naming a mapping record type. Valid values are
+      #               :salesforce_record_type or :database_record_type.
+      # options     - A Hash of options to pass to `each` (optional).
+      #
+      # Returns an Array of Restforce::DB::Instances::Base.
+      def collection(mapping, record_type, options = {})
+        if cached?(mapping, record_type)
```

Comment:
I wouldn't mind an early `return` from this method to decrease the indentation level below.